server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -77,10 +89,27 @@
 
     private AggregatedDfs aggregatedDfs;
     private QueryProfiler profiler;
-    private Runnable checkCancelled;
+    private Holder<Cancellable> cancellable;
```

Comment:
why do you need to use a custom `Holder` ? Now that cancellable directory reader is forked, you should be able to set the cancellable lazily ? So instead of passing the `Cancellable` here you'd set it in `ContextIndexSearcher#setCancellable` ?

Reply:
The issue is that we need a reference to the cancellable (the holder actually) during construction of the Reader: https://github.com/elastic/elasticsearch/pull/52822/files#diff-913ad694a7d744ee93ae1dac48d67b0eR403
So that later on with the `setCancellable` we can set it to the holder.

Reply:
Alreanatively I changed so that the ContextIndexSearcher doesn't know of the holder, but then a cast to `CancellableDirectoryReader` is required to call the setCancellable.